### PR TITLE
Ability for Hash::sort to sort case-insensitively

### PR DESCRIPTION
**Pull request by aygee**

### CODE REVIEW

**Comment by dereuromark** [comment]
At some point we sure have to refactor it into an options array like

    public static function sort(array $data, $path, array $options = array()) {}

File: lib/Cake/Utility/Hash.php [hunk 1]
```diff
@@ -870,9 +871,16 @@ public static function sort(array $data, $path, $dir = 'asc', $type = 'regular')
 
 		$dir = strtolower($dir);
 		$type = strtolower($type);
-		if ($type === 'natural' && version_compare(PHP_VERSION, '5.4.0', '<')) {
-			$type = 'regular';
+
+		// Natural and case insensitive sort is only supported from >= 5.4.0
+		if (version_compare(PHP_VERSION, '5.4.0', '<')) {
+			if ($type === 'natural' || $type === 'natural_ignore_case' || $type === 'regular_ignore_case') {
```
Comment (markstory) [inline lib/Cake/Utility/Hash.php]: You have a boolean option and string types here.

File: lib/Cake/Utility/Hash.php [hunk 2]
```diff
@@ -843,10 +843,11 @@ public static function apply(array $data, $path, $function) {
  * @param string $path A Set-compatible path to the array value
  * @param string $dir See directions above. Defaults to 'asc'.
  * @param string $type See direction types above. Defaults to 'regular'.
+ * @param string $ignoreCase Case insensitive sorting. Defaults to false.
  * @return array Sorted array of data
  * @link http://book.cakephp.org/2.0/en/core-utility-libraries/hash.html#Hash::sort
  */
-	public static function sort(array $data, $path, $dir = 'asc', $type = 'regular') {
+	public static function sort(array $data, $path, $dir = 'asc', $type = 'regular', $ignoreCase = false) {
```
Comment (markstory) [inline lib/Cake/Utility/Hash.php]: I'm not a fan of boolean flags. They make method calls harder to understand at a glance. Could we either overload $type into an options array, or have more types?

Something like

```php
if (is_string($type)) {
  $type = ['type' => $type];
}
$type = (array)$type + ['type' => 'regular', 'caseSensitive' => true];
```
could be one way to overload the parameter.

File: lib/Cake/Test/Case/Utility/HashTest.php [hunk 3]
```diff
@@ -1388,6 +1425,50 @@ public function testSortString() {
...
+		);
+		$sorted = Hash::sort($toSort, '{n}.Item.name', 'asc', 'regular', true);
+		$expected = array(
+			array('Item' => array('name' => 'Baby')),
+			array('Item' => array('name' => 'bar')),
+			array('Item' => array('name' => 'bat')),
+			array('Item' => array('name' => 'Baz')),
+		);
+		$this->assertEquals($expected, $sorted);
+	}
```
Comment (markstory) [inline lib/Cake/Test/Case/Utility/HashTest.php]: Test cases look good.

File: lib/Cake/Utility/Hash.php (hunk 2 above)
Comment (aygee) [inline lib/Cake/Utility/Hash.php]: I like this solution. Will implement it this way. Thanks!

File: lib/Cake/Utility/Hash.php (hunk 1 above)
Comment (aygee) [inline lib/Cake/Utility/Hash.php]: Sorry, this was my first attempt to solve it, but didn't clean it properly. Will clean it up. Thanks.

**Comment by aygee** [comment]
Updated the code, not sure why the ci failed. I don't think its related to my code

**Comment by ADmad** [comment]
This is still one coding standard [failure](https://travis-ci.org/cakephp/cakephp/jobs/75552131#L234).

**Comment by aygee** [comment]
I think I have removed the whitespace at the end of line Hash.php but CI still failed. I didn't actually add that whitespace, it was already there.

    FILE: /home/travis/build/cakephp/cakephp/lib/Cake/Utility/Hash.php
    --------------------------------------------------------------------------------
    FOUND 1 ERROR(S) AFFECTING 1 LINE(S)
    --------------------------------------------------------------------------------
     876 | ERROR | Whitespace found at end of line

**Comment by ADmad** [comment]
> I didn't actually add that whitespace, it was already there.

So just remove it :)

File: lib/Cake/Utility/Hash.php [hunk 4]
```diff
@@ -869,10 +872,30 @@ public static function sort(array $data, $path, $dir = 'asc', $type = 'regular')
...
+			if (!empty($type['ignoreCase'])) {
+				$type = $type['type'];
+			} else {
+				throw new InvalidArgumentException(__d('cake_dev',
+					'Invalid parameter $type. It requires type key to be specified.'
+				));
+			}
+		} else {
+			$type = strtolower($type);
+		}
```
Comment (markstory) [inline lib/Cake/Utility/Hash.php]: This feels like a lot of branching code that could probably be avoided by merging defaults.

File: lib/Cake/Utility/Hash.php [hunk 5]
```diff
@@ -869,10 +872,30 @@ public static function sort(array $data, $path, $dir = 'asc', $type = 'regular')
...
+		$ignoreCase = false;
+		
+		// $type can be overloaded for case insensitive sort
+		if (is_array($type)) {
+
+			if (!empty($type['ignoreCase'])) {
+				$ignoreCase = $type['ignoreCase'];
+			}
+
+			if (!empty($type['ignoreCase'])) {
```
Comment (markstory) [inline lib/Cake/Utility/Hash.php]: Why does one key infer the other?

**Comment by markstory** [comment]
Thanks @aygee. Merged into 2.8

**Comment by aygee** [comment]
@markstory, thanks for fixing the code. Sorry I was going to do it at some point but time got the better of me. Cheers.

**Pull request merged**